parser-typechecker/src/Unison/Codebase/Causal.hs

```diff
@@ -297,14 +136,25 @@ stepDistinctM
   => (e -> n e) -> Causal m h e -> n (Causal m h e)
 stepDistinctM f c = (`consDistinct` c) <$> f (head c)
 
+-- | Causal construction should go through here for uniformity;
+-- with an exception for `one`, which avoids an Applicative constraint.
+fromList :: (Applicative m, Hashable e) => e -> [Causal m h e] -> Causal m h e
+fromList e (List.nubOrdOn currentHash -> tails) = case tails of
+  [] -> UnsafeOne h e
+  t : [] -> UnsafeCons h e (tailPair t)
```

Comment:
What should we do here if `h` is equal to `currentHash t`?
We should probably de-dupe this right?

Reply:
No, we do that in `consDistinct`, but not in `cons`.